providers/keycloak.go

```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
 	p.Group = group
 }
 
-func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
-	json, err := requests.New(p.ValidateURL.String()).
+func (p *KeycloakProvider) SetRoles(roles []string) {
+	p.Roles = roles
+}
+
+func extractRolesFromClaims(claims map[string]interface{}) ([]string, error) {
+	var roleList []string
+
+	if realmRoles, found := claims["realm_access"].(map[string]interface{}); found {
```

Comment:
By chance do you have a sample of what Keycloak IDToken claims are present? I'm curious if some of the data pulled via the userinfo/profileURL endpoint can potentially be extracted from IDToken claims and we don't need that extra HTTP call.

Reply:
(This is definitely scope creep and not needed for your PR -- I'm just curious for my own understanding).

Reply:
Not sure about the IDToken, but the AccessToken contains both the email and the groups which should be sufficient to get the info used here. I would say that HTTP call is not particularly necessary.

Reply:
Although that might require a breaking change since I believe that's configurable, if groups are returned via userinfo but not included in the token.

Reply:
I checked, and it should possible to get the required info from the IDToken as well with the right configuration.

Reply:
Oh interesting... that would definitely make this different than other providers that are more OIDC based and get extra claims from an IDToken we can verify.

Does Keycloak have any documentation on where to get the public key to verify the signature on the Access Tokens?

I really only want to use claims extracted from an unverified token as a last resort. There's gotta be a way we can verify the signature of the token?

Reply:
Is this an option? https://stackoverflow.com/questions/54318633/getting-keycloaks-public-key

Reply:
Keycloak has a JWKS endpoint at `/auth/realms/<realm_name>/protocol/openid-connect/certs`, it should be possible to verify it using that, should I reuse `oidc-jwks-url` for this purpose? If so, what would be the best way to pass it to the provider?

Reply:
So I need your help explaining something 😅 

Keycloak as a provider confuses me. When I look at it, it has barebones implementations as if it were vanilla OAuth2. (and should what it calls the `validateURL` be the `profileURL`...).

BUT - I get sample configs from users using OIDC config options with it set as the provider in questions frequently. If it is truly OIDC compliant in all cases, I feel like it should be baked into the provider more.

This way we can take advantage of all the new OIDC based features we are making global and Keycloak can take advantage of them (bearer token to session logic, generic claims to session fields - future adhoc claims -> session -> header support).

The reason Keycloak would still be a distinct provider instead of just using Keycloak with an OIDC provider: your custom roles from access tokens logic here, any special quirks in the schema of the userinfo endpoint etc.

Global OIDC logic refactor: https://github.com/oauth2-proxy/oauth2-proxy/pull/936

CC: @JoelSpeed -- if you have some history context or experience with this provider yourself, I would love your thoughts.

Reply:
My understanding is that it is OIDC-compliant (it's OpenID-certified: https://openid.net/developers/certified/), I believe validateURL should be userInfoURL.
It _should_ work with the standard OIDC provider implementation (although I have not tried that myself), but yeah basically the additional role functionality is probably worth having it as an independent provider.

Reply:
So I think to get at an OIDC IDToken Verifier (if that works on the Access Token) or underlying raw JWKS, we use the existing flags to set it: https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/oauth_provider/#skip-oidc-discovery

During options validation, if you see `--keycloak-roles` you'd need to assert you also have the OIDC URLs setup so you have the tools at your disposal to verify the signature of the Access Token.

Reply:
And on the userinfo... I guess technically it works as a validateURL since it will take an access token and return 200, which is all this is looking for: https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/providers/internal_util.go#L49

But it is primarily used as a profileURL in the Keycloak provider itself.  I'll see if I can get a PR up to clean that up in a backwards compatible way.